aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -0,0 +1,514 @@
+  //  torch.tensor(3, dtype=torch.uint8) + 5 -> torch.uint8
+  //
+  auto result_type = ScalarType::Undefined;
+  auto backend = Backend::Undefined;
+  std::tie(result_type, backend) = compute_result_type(operands_, [](const Tensor& t) {
+    return t.dim() > 0;
+  });
+  if (result_type == ScalarType::Undefined) {
+    std::tie(result_type, backend) = compute_result_type(operands_, [](const Tensor& t) {
+      return !t.get()->is_wrapped_number();
+    });
+  }
+  if (result_type == ScalarType::Undefined) {
+    std::tie(result_type, backend) = compute_result_type(operands_, [](const Tensor& t) {
+      return true;
+    });
+  }
+
+  AT_ASSERT(result_type != ScalarType::Undefined);
+  AT_ASSERT(backend != Backend::Undefined);
+
+  auto& type = at::globalContext().getType(backend, result_type);
+
+  for (auto& op : operands_) {
+    if (!op.type_) {
+      op.type_ = &type;
+      if (op.tensor_->defined() && type != op.tensor_->type()) {
+        if (op.tensor_->dim() == 0) {
+          if (type.backend() != at::kCUDA) {
+            *op.tensor_ = op.tensor_->toType(type);
+          }
+        } else {
+          op.needs_cast_ = true;
+        }
+      }
+    }
+  }
+}
+
+DimVector TensorIterator::compatible_stride(int element_size) const {
+  auto stride = DimVector();
+  stride.push_back(element_size);
+  for (int i = 0; i < ndim() - 1; i++) {
+    stride.push_back(shape_[i] * stride[i]);
+  }
+  return stride;
+}
+
+DimVector TensorIterator::invert_perm(IntList input) const {
+  // Invert the permutation caused by reorder_dimensions. This is not valid
+  // after coalesce_dimensions is called.
+  auto res = DimVector(input.size(), 0);
+  for (int dim = 0; dim < ndim(); dim++) {
+    res[perm_[dim]] = input[dim];
+  }
+  return res;
+}
+
+void TensorIterator::allocate_outputs() {
+  for (int i = 0; i < num_outputs_; i++) {
+    auto& op = operands_[i];
+    if (!op.tensor_->defined()) {
+      int element_size = op.type_->elementSizeInBytes();
+      op.stride_ = compatible_stride(element_size);
+
+      auto tensor_shape = invert_perm(shape_);
+      auto tensor_stride = invert_perm(op.stride_);
+      for (int dim = 0; dim < ndim(); dim++) {
+        tensor_stride[dim] /= element_size;
+      }
+      *op.tensor_ = op.type_->tensor(tensor_shape, tensor_stride);
+    }
+  }
+}
+
+void TensorIterator::coalesce_dimensions() {
+  if (ndim() == 0) {
+    return;
+  }
+
+  // We can coalesce two adjacent dimensions if either dim has size 1 or if:
+  // shape[n] * stride[n] == shape[n + 1].
+  auto can_coalesce = [&](int dim0, int dim1) {
+    auto shape0 = shape_[dim0];
+    auto shape1 = shape_[dim1];
+    if (shape0 == 1 || shape1 == 1) {
+      return true;
+    }
+    for (int i = 0; i < ntensors(); i++) {
+      auto& stride = operands_[i].stride_;
+      if (shape0 * stride[dim0] != stride[dim1]) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  auto copy_strides = [&](int dim0, int dim1) {
```

Comment:
nit: I'd call this coalesec_stride_dim or something -- 'copying' makes me think of a different op.